---
title: "Class 9: Structural Bioinformatics 1"
author: "Anh Tran"
format: html
---

To analyze PDB file, can download csv and use read.csv() to analyze the data. Protein structures by X-ray crystalography dominate this database. 
We are skipping Q1-3 as the website was too slow for us.

```{r}
d <- read.csv("Data Export Summary.csv")
head(d)
```

>Q1: What percentage of structures in the PDB are solved by X-Ray and Electron Microscopy.

```{r}
(169794 + 13825)/196779
```

>Q2: What proportion of structures in the PDB are protein?

```{r}
(171221 + 10444 + 10876)/196779
```

>Q3: Type HIV in the PDB website search box on the home page and determine how many HIV-1 protease structures are in the current PDB?

2,545 Structures

![HIV-Pr structure from 1hsg](1HSG.png)

>Q4: Water molecules normally have 3 atoms. Why do we see just one atom per water molecule in this structure?

Because the resolution is not high enough/hydrogen molecules are too small so only oxygen atom is displayed per water molecule. 

> Q5: There is a critical “conserved” water molecule in the binding site. Can you identify this water molecule? What residue number does this water molecule have

Yes, 308

>Q6: Generate and save a figure clearly showing the two distinct chains of HIV-protease along with the ligand. You might also consider showing the catalytic residues ASP 25 in each chain (we recommend “Ball & Stick” for these side-chains). Add this figure to your Quarto document.
Discussion Topic: Can you think of a way in which indinavir, or even larger ligands and substrates, could enter the binding site?

![HIV-Pr structure from 1hsg](1HSG.custom.png)




## 3. Introduction to Bio3D in R
Bio3D is an R packaged for structural bioinformatcis. To use it we ned to call it up with the `library` function (just like any package)

```{r}
library(bio3d)
```

To read PDF file we can use `read.pdb()`

```{r}
pdb <- read.pdb("1hsg")
pdb
```

>Q7: How many amino acid residues are there in this pdb object? 

198

>Q8: Name one of the two non-protein residues? 

MK1

>Q9: How many protein chains are in this structure? 

2

```{r}
attributes(pdb)
```

The ATOM records of a PDB file are stored in `pdb$atom`

```{r}
head(pdb$atom)
```

```{r}
adk <- read.pdb("6s36")
```

```{r}
adk
```

```{r}
m <- nma(adk)
```

```{r}
plot(m)
```

```{r}
mktrj(m, file="adk_m7.pdb")
```

# 4. Comparative analysis of Adenylate kinase (ADK)

We will start our analysis with a single PDB id (code from the PDB database): 1AKE

First we get its primary sequence:

```{r}
#install.packages("bio3d")
#install.packages("devtools")
#install.packages("BiocManager")
#BiocManager::install("msa")
#devtools::install_bitbucket("Grantlab/bio3d-view")
```

>Q10. Which of the packages above is found only on BioConductor and not CRAN?

msa

>Q11. Which of the above packages is not found on BioConductor or CRAN?

bio3d-view

>Q12. True or False? Functions from the devtools package can be used to install packages from GitHub and BitBucket?

TRUE

```{r}
aa <- get.seq("1ake_a")
aa
```

>Q13. How many amino acids are in this sequence, i.e. how long is this sequence? 

214

```{r}
# Blast or hmmer search 
#b <- blast.pdb(aa)
```

```{r}
# Plot a summary of search results
#hits <- plot(b)
```

```{r}
# List out some 'top hits'
hits <- NULL
hits$pdb.id <- c('1AKE_A','6S36_A','6RZE_A','3HPR_A','1E4V_A','5EJE_A','1E4Y_A','3X2S_A','6HAP_A','6HAM_A','4K46_A','3GMT_A','4PZL_A')
head(hits$pdb.id)
```

```{r}
# Download releated PDB files
files <- get.pdb(hits$pdb.id, path="pdbs", split=TRUE, gzip=TRUE)
```

```{r}
# Align releated PDBs
pdbs <- pdbaln(files, fit = TRUE, exefile="msa")
```

```{r}
# Vector containing PDB codes for figure axis
ids <- basename.pdb(pdbs$id)

# Draw schematic alignment
plot(pdbs, labels=ids)
```

```{r}
anno <- pdb.annotate(ids)
unique(anno$source)
```

```{r}
anno
```

# Jump to PCA

```{r}
# Perform PCA
pc.xray <- pca(pdbs)
plot(pc.xray)
```

```{r}
# Calculate RMSD
rd <- rmsd(pdbs)

# Structure-based clustering
hc.rd <- hclust(dist(rd))
grps.rd <- cutree(hc.rd, k=3)

plot(pc.xray, 1:2, col="grey50", bg=grps.rd, pch=21, cex=1)
```
## 6. Normal mode analysis [optional]

```{r}
# NMA of all structures
modes <- nma(pdbs)
```

```{r}
plot(modes, pdbs, col=grps.rd)
```

